crates/libs/windows/src/core/heap.rs

```diff
@@ -32,20 +32,43 @@ pub unsafe fn heap_free(ptr: RawPtr) {
     }
 }
 
-/// Copy a slice of `T` into a freshly allocated buffer with an additional default `T` at the end.
+/// Copy an iterator of `T` into a freshly allocated buffer with an additional default `T` at the end.
 ///
-/// Returns a pointer to the beginning of the buffer
+/// Returns a pointer to the beginning of the buffer. This pointer must be freed when done using `heap_free`.
 ///
 /// # Panics
 ///
 /// This function panics if the heap allocation fails or if the pointer returned from
 /// the heap allocation is not properly aligned to `T`.
-pub fn heap_string<T: Copy + Default + Sized>(slice: &[T]) -> *const T {
-    unsafe {
-        let buffer = heap_alloc((slice.len() + 1) * std::mem::size_of::<T>()).expect("could not allocate string") as *mut T;
-        assert!(buffer.align_offset(std::mem::align_of::<T>()) == 0, "heap allocated buffer is not properly aligned");
-        buffer.copy_from_nonoverlapping(slice.as_ptr(), slice.len());
-        buffer.add(slice.len()).write(T::default());
-        buffer
+///
+/// # Safety
+/// len must not be less than the number of items in the iterator.
+pub unsafe fn string_from_iter<I, T>(iter: I, len: usize) -> *const T
+where
+    I: Iterator<Item = T>,
+    T: Copy + Default,
+{
+    let str_len = len + 1;
+    let ptr = heap_alloc(str_len * std::mem::size_of::<T>()).expect("could not allocate string") as *mut T;
+
+    // TODO this assert is mostly redundant, HeapAlloc has alignment of 8, we currently only require alignments of 1 or 2.
+    // There is no meaningful string type with characters that require an alignment above 8.
+    assert_eq!(ptr.align_offset(std::mem::align_of::<T>()), 0, "heap allocated buffer is not properly aligned");
+
+    let mut encoder = iter.chain(core::iter::once(T::default()));
+
+    for i in 0..str_len {
+        core::ptr::write(
+            ptr.add(i),
+            match encoder.next() {
+                Some(encoded) => encoded,
+                None => break,
+            },
+        );
     }
+
+    // TODO ensure `encoder` is a fused iterator
+    assert!(encoder.next().is_none(), "encoder returned more characters than expected");
```

Comment:
Won't this be fused because the chaining of `once()` is guaranteed to be fused?  `iter` might not be fused, but the implementation will never call `next()` on it again because the chain moved on to the next one.

Reply:
Nice catch, I didn't consider that. `Once` is indeed guaranteed to be fused, so that issue is solved!

Reply:
`asserting` here seems wrong... I'm not sure we want to panic if the encoder contains more than `len`. If anything, a `debug_assert!` might be appropriate (though I think silently ignoring is likely ok).

Reply:
I'd also like to avoid panics in general. I have customers for whom this is inappropriate. If it means we have to embed this inside `PCWSTR` and `PCSTR` directly to avoid the issues with generality, then so be it.